src/daemon/rpmostreed-transaction-livefs.c

```diff
@@ -522,6 +522,106 @@ replace_subpath (OstreeRepo *repo,
   return TRUE;
 }
 
+/* The sledgehammer 🔨 approach. Because /usr is a mount point, we can't replace
+ * all of it. We could do a diff, but doing that precisely and quickly depends
+ * on https://github.com/ostreedev/ostree/issues/1224
+ *
+ * In this approach, we iterate over and exchange subdirectories. Some issues
+ * here are processes that hold directory fds open will have those invalidated,
+ * even if they shouldn't otherwise be affected. Another issue is that if the
+ * kernel is too old to have `RENAME_EXCHANGE`, we'll have e.g. `/usr/bin` be
+ * temporarily broken.
+ *
+ * Yet another issue is that doing things all at once makes it much more likely
+ * that we'll e.g. replace code for a program before updating a shared library
+ * it depends on. There's really no fixing that problem in general, but we could
+ * minimize the race window in the same way package systems tend to do by doing
+ * the filesystem tree replacements in package reverse dependency order.
+ *
+ * On the other hand, this handles tricky cases like replacing a directory with
+ * a regfile or symlink.
+ *
+ */
+static gboolean
+replace_usr (OstreeRepo *repo,
+             int deployment_dfd,
+             GLnxTmpDir *tmpdir,
+             CommitDiff *diff,
+             const char *target_csum,
+             GCancellable *cancellable,
+             GError **error)
+{
+  /* Grab a reference to the current /usr */
+  glnx_fd_close int deployment_usr_dfd = -1;
+  if (!glnx_opendirat (deployment_dfd, "usr", TRUE, &deployment_usr_dfd, error))
+    return FALSE;
+
+  /* Check out our new /usr */
+  OstreeRepoCheckoutAtOptions usr_checkout_opts = { .mode = OSTREE_REPO_CHECKOUT_MODE_NONE,
+                                                    .overwrite_mode = OSTREE_REPO_CHECKOUT_OVERWRITE_NONE,
+                                                    .no_copy_fallback = TRUE,
+                                                    .subpath = "/usr" };
+
+  if (!ostree_repo_checkout_at (repo, &usr_checkout_opts, tmpdir->fd, "usr",
+                                target_csum, cancellable, error))
+    return FALSE;
+
+  /* Iterate over new /usr, see whether the context exists. If not, just
+   * rename(). If so, `RENAME_EXCHANGE`; the old content will be rm-rf'd since
+   * it will be moved to the tmpdir.
+   */
+  g_autoptr(GHashTable) seen_new_children = g_hash_table_new (g_str_hash, g_str_equal);
+  g_auto(GLnxDirFdIterator) dfd_iter = { FALSE, };
+  if (!glnx_dirfd_iterator_init_at (tmpdir->fd, "usr", TRUE, &dfd_iter, error))
+    return FALSE;
+  while (TRUE)
+    {
+      struct dirent *dent = NULL;
+      if (!glnx_dirfd_iterator_next_dent (&dfd_iter, &dent, cancellable, error))
+        return FALSE;
+      if (dent == NULL)
+        break;
+      const char *name = dent->d_name;
+      /* Keep track of what entries are in the new /usr */
+      g_hash_table_add (seen_new_children, g_strdup (name));
```

Comment:
We're leaking this, right? Hash table should probably be `g_hash_table_new_full (..., ..., g_free, NULL)`